File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution{
public:
    int lengthOfLongestSubstring(string s) {
        int i = 0, j = 0, n = s.length(), len = 0;
        unordered_set<char> st;

        while (j < n) {
            if (st.find(s[j]) == st.end()) {
                st.insert(s[j++]);
                len = max((int)st.size(), len);
            }
            else st.erase(s[i++]);
        }

        return len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int i = 0, j = 0, maxLen = 0;
        Set<Character> set = new HashSet<>();

        while (j < n) {
            char currChar = s.charAt(j);
            if (!set.contains(currChar)) {
                set.add(currChar);
                maxLen = Math.max(maxLen, j - i + 1);
                j++; // Only move right pointer when we successfully expand
            } else {
                set.remove(s.charAt(i));
                i++; // Shrink window from left until duplicate is gone
            }
        }
        return maxLen;
    }
}